feat(#8204): Add feature to support adding phone number when registering patients

PrjShrestha opened the pull request:
Feature [(#8204)](https://github.com/medic/cht-core/issues/8204): Support adding phone number when registering patients via SMS


# Description
This code aims to complete the following: 
Users should be able to add a patient's phone number while registering a patient.
CHT should automatically append the default country code to the phone number before saving it to the database.

Sample Message: 
![image](https://github.com/medic/cht-core/assets/3299006/539f9718-87ba-4819-8972-d5e4aea448ab)
****

[x]  Added unit tests
[x] Tested locally by setting up cht
[x]  Fixed eslint issues
[x] [Doc update for new field phone_number](https://github.com/medic/cht-docs/pull/1152)
[x] [Doc update validation] (https://github.com/medic/cht-docs/pull/1144)

binokaryg commented on config/default/app_settings.json:
```diff
@@ -1,454 +1,3205 @@
 {
-  "languages": [
```
Why are there so many changes in this file?

PrjShrestha replied:
Accidentally committed settings created for testing. Reverted.

binokaryg commented on shared-libs/transitions/src/transitions/registration.js:
```diff
@@ -185,6 +186,27 @@ const setBirthDate = doc => {
+  const validPhone = phoneNumberParser.validate(config.getAll(), phoneNumber);
+  if (!validPhone) {
+    throw new Error(
+      `${doc.phoneNumber} submitted by ${doc.from} is not a valid phone number`
```
Can we route this to an event and allow a custom message to be sent similar to the `patient_not_found` event?
https://github.com/medic/cht-core/blob/1d19531b2404595d17f63b631c659a717a786ee5/tests/integration/sentinel/transitions/update-notifications.spec.js#L116

binokaryg replied:
Also, what if we don't want to allow duplicate phone numbers? What if others want to allow it? Can that be made configurable?

PrjShrestha replied:
It seems we can. Thanks. I added a phone not valid event and its translations (en , ne).

binokaryg commented on shared-libs/transitions/src/transitions/registration.js:
```diff
@@ -185,6 +186,27 @@ const setBirthDate = doc => {
+  const phoneNumber = getPhoneNumber(doc);
+  // By default for a valid phone number SmsParser itself adds a country code if not provided
+  // So if country code is not present in phone_number throw error
+  const validPhone = phoneNumberParser.validate(config.getAll(), phoneNumber);
```
Also, what if we don't want to allow duplicate phone numbers (as in contact forms)? What if others want to allow duplicates? Can that be made configurable?

PrjShrestha replied:
Now its configurable. Added a validation uniquePhone for the phone_number field which controls this.

garethbowen commented on tests/integration/sentinel/transitions/registration.spec.js:
```diff
@@ -98,7 +97,225 @@ describe('registration', () => {
   afterEach(() => utils.revertDb(getIds(contacts), true));
 
-  it('should be skipped when transition is disabled', () => {
+  it.only('should add valid phone to patient doc', () => {
```
Please revert the `.only` so all tests run in CI.

PrjShrestha replied:
Thanks. Done.

garethbowen commented on tests/integration/sentinel/transitions/registration.spec.js:
```diff
@@ -98,7 +97,225 @@ describe('registration', () => {
+      });
+  });
+
+  it(' be skipped when transition is disabled', () => {
```
Looks like you modified the test name accidentally?

```suggestion
  it('should be skipped when transition is disabled', () => {
```

garethbowen reviewed (changes requested):
Nice one. A couple of changes requested...

yrimal commented on api/resources/translations/messages-ne.properties:
```diff
@@ -764,7 +764,8 @@ messages.errors.patient.missing =
 messages.f.report_accepted_parent = 
 messages.generic.no_provided_patient_id = 
-messages.generic.provided_patient_id_not_unique = 
+messages.generic.provided_patient_id_not_unique =
```
Please add translation for this.

PrjShrestha replied:
Done.

yrimal commented on api/resources/translations/messages-ne.properties:
```diff
@@ -764,7 +764,8 @@ messages.errors.patient.missing =
 messages.generic.no_provided_patient_id = 
-messages.generic.provided_patient_id_not_unique = 
+messages.generic.provided_patient_id_not_unique =
+messages.generic.provided_phone_not_valid = तपैले पठाउनुभएको फोन नम्बर मिलेन । फेरी प्रयास गर्नुहोस्।
```
```suggestion
messages.generic.provided_phone_not_valid =तपाइँले पठाउनुभएको फोन नम्बर मिलेन। फेरी प्रयास गर्नुहोस्।
```

PrjShrestha replied:
Pardon my Nepali :). Done.

yrimal commented on shared-libs/validation/test/validations.js:
```diff
@@ -248,6 +248,61 @@ describe('validations', () => {
     });
   });
 
+  it.only('should fail uniquePhone if doc is found in db', () => {
```
Did you mean to check `phone` uniqueness here? If so can you format the test name like: 
`should fail uniquePhone if phone is found in db`

PrjShrestha replied:
Thanks. Done.

yrimal commented on shared-libs/validation/test/validations.js:
```diff
@@ -248,6 +248,61 @@ describe('validations', () => {
+    });
+  });
+
+  it.only('should pass uniquePhone if doc is not found in db', () => {
```
I assume you meant to check phone and not doc here.

PrjShrestha replied:
Was checking if a contact doc has the same phone. Accepted suggestion to rename. Makes more sense.

yrimal commented on api/resources/translations/messages-ne.properties:
```diff
@@ -764,7 +764,8 @@ messages.errors.patient.missing =
 messages.f.report_accepted_parent = 
 messages.generic.no_provided_patient_id = 
-messages.generic.provided_patient_id_not_unique = 
+messages.generic.provided_patient_id_not_unique = तपाइँले पठाउनुभएको आइडी पैलै प्रयोग मा छ।  
```
```suggestion
messages.generic.provided_patient_id_not_unique = तपाइँले पठाउनुभएको आई.डी पहिल्यै प्रयोगमा छ।  
```

PrjShrestha replied:
Changed. Thanks.

yrimal commented on api/tests/mocha/services/report/smsparser.spec.js:
```diff
@@ -97,10 +97,82 @@ describe('sms parser', () => {
     chai.expect(smsparser.getFormCode('द CDT33')).to.equal('द');
   });
 
+  it('accepts form with phone_number type', () => {
+    const doc = { message: 'NP 20 9841202020' };
+    const def = definitions.forms.NP;
+    sinon.stub(config, 'getAll').returns({
+      default_country_code: 977,
+      phone_validation: 'full'
+    });
+    const data = smsparser.parse(def, doc);
+    chai.expect(data.phone_number).to.equal('+9779841202020');
+  });
+
+  it('accepts phone number with extension', () => {
+    const doc = { message: 'NP 20 +9779841202020' };
+    const def = definitions.forms.NP;
+    sinon.stub(config, 'getAll').returns({
+      default_country_code: 977,
+      phone_validation: 'full'
+    });
+    const data = smsparser.parse(def, doc);
+    chai.expect(data.phone_number).to.equal('+9779841202020');
+  });
+
+  it('accepts correct phone number without extension', () => {
+    const doc = { message: 'NP 20 9841202020' };
+    const def = definitions.forms.NP;
+    sinon.stub(config, 'getAll').returns({
+      default_country_code: 977,
+      phone_validation: 'full'
+    });
+    const data = smsparser.parse(def, doc);
+    chai.expect(data.phone_number).to.equal('+9779841202020');
```
Isn't 3rd test the exact duplicate of first test here ?

PrjShrestha replied:
Hmm. Yes. Any ideas on how I can distinguish testing phone_number field is supported i an different way or the second test should suffice?

@yrimal

garethbowen commented on api/src/services/report/smsparser.js:
```diff
@@ -209,9 +210,19 @@ const fieldParsers = {
+    if (formattedAndValidatedPhone) {
+      return formattedAndValidatedPhone;
+    } else {
+      logger.error(`The provided phone number ${raw} is invalid`);
```
Other validation failures are logged as `warn` - I think this is because it's not a failure of the code. I think we should keep this one consistent with those.

```suggestion
      logger.warn(`The provided phone number ${raw} is invalid`);
```

garethbowen commented on api/tests/mocha/services/report/smsparser.spec.js:
```diff
@@ -97,10 +97,71 @@ describe('sms parser', () => {
+  [['NP 20 +918750660880 Prajwol', '+918750660880'],
+    ['NP 20 +254773087889 Prajwol', '+254773087889'],
+    ['NP 20 +255712262987 Prajwol', '+255712262987']].forEach(phoneNumerWithParsed => {
+    it('returns parsed number if valid phone of another the region', () => {
```
This will list three unit tests with identical names. If one of them fails it won't be immediately obvious which one. Let's include the message in the test name:

```suggestion
    it(`returns parsed number if valid phone of another the region: ${phoneNumerWithParsed[0]}`, () => {
```

garethbowen commented on shared-libs/transitions/src/transitions/utils.js:
```diff
@@ -58,7 +58,12 @@ module.exports = {
+      .query('medic-client/contacts_by_reference', { key: ['shortcode', id] })
+      .then(results => !(results && results.rows && results.rows.length));
+  },
+  isPhoneUnique: (phoneNumber) => {
```
It looks like the `transitions/utils` file already imports '@medic/validation' so if you expose the new `uniquePhone` function there we can reuse this code.

PrjShrestha replied:
@garethbowen The code in utils was redundant hence I removed it. Could have been done by putting the duplicate check in utils but couldn't find any way to stop all further transitions when phone was duplicate.

garethbowen commented on tests/integration/sentinel/transitions/registration.spec.js:
```diff
@@ -98,6 +97,225 @@ describe('registration', () => {
+        }],
+      }],
+      forms: { 'FORM-A': {} }
+    };
```
There's a lot of repetition in this file, eg: the `const settings` are identical. Let's pull that out as a constant (`PHONE_NUMBER_VALIDATION_CONFIG`), or a function (`getPhoneNumberValidationConfig`) to make that clear.

PrjShrestha replied:
@garethbowen Thanks. Created a stub-form file where the form(settings) is not read from.

garethbowen reviewed (changes requested):
Thank you, this is awesome! I've left a few comments for cleanup but nothing major.

Also, a recent change in the required builds means that to get a green build you'll now need to rebase to head to get the new test config. Sorry about that!

garethbowen commented on config/default/app_settings.json:
```diff
@@ -353,9 +353,9 @@
   "tasks": {
-    "rules": "define Target { _id: null, contact: null, deleted: null, type: null, pass: null, date: null, groupBy: null }\ndefine Contact { contact: null, reports: null, tasks: null }\ndefine Task { _id: null, deleted: null, doc: null, contact: null, icon: null, date: null, readyStart: null, readyEnd: null, title: null, fields: null, resolved: null, priority: null, priorityLabel: null, reports:  ...
+    "rules": "define Target { _id: null, contact: null, deleted: null, type: null, pass: null, date: null, groupBy: null }\ndefine Contact { contact: null, reports: null, tasks: null }\ndefine Task { _id: null, deleted: null, doc: null, contact: null, icon: null, date: null, readyStart: null, readyEnd: null, title: null, fields: null, resolved: null, priority: null, priorityLabel: null, reports:  ...
     "targets": {
```
Can you tell me what changed in this file? This config is used as the configuration when the app is first started so we have to be a little careful changing it, and it's hard to diff because it's minified...

PrjShrestha replied:
Nothing should have changed. Must be formatting changes i will revert.

garethbowen commented on shared-libs/transitions/src/transitions/registration.js:
```diff
@@ -185,6 +186,27 @@ const setBirthDate = doc => {
+  const phoneNumber = getPhoneNumber(doc);
+  // By default for a valid phone number SmsParser itself adds a country code if not provided
+  // So if country code is not present in phone_number throw error
+  const app_settings = config.getAll();
+  const validPhone = phoneNumberParser.validate(app_settings, phoneNumber);
```
```suggestion
  const appSettings = config.getAll();
  const validPhone = phoneNumberParser.validate(appSettings, phoneNumber);
```

Nitpick: Just a minor style guide change - JS variables should be camel case.

garethbowen reviewed (changes requested):
This looks great! One comment and one very minor change requested.

I'll get a QA to help to assist in testing and then it should be good to go!

binokaryg reviewed (changes requested):
Minor changes requested.

PrjShrestha commented on api/tests/form-definitions.js:
```diff
@@ -791,5 +791,75 @@ exports.forms = {
+          1,
+          3
+        ],
+        'range': [
+          0,
+          125
+        ],
```
Removed those validations.

PrjShrestha commented on api/tests/form-definitions.js:
```diff
@@ -791,5 +791,75 @@ exports.forms = {
       }
     }
+  },
+  'NP': {
```
Done

binokaryg replied:
If it's only being used for `api/tests/mocha/services/report/smsparser.spec.js`, tests are passing with just this:
```
  NP: {
    meta: {
      code: 'NP'
    },
    fields: {
      patient_age: {
        type: 'integer'
      },
      phone_number: {
        type: 'phone_number'
      }
    }
  }
```

PrjShrestha replied:
I had originially kept that form to look like how a real form would be but as its being used for tests only I will change it.

PrjShrestha commented on api/tests/mocha/services/report/smsparser.spec.js:
```diff
@@ -97,10 +97,71 @@ describe('sms parser', () => {
+    const def = definitions.forms.NP;
+    sinon.stub(config, 'getAll').returns({
+      default_country_code: 977,
+      phone_validation: 'full'
```
Added test to check all validation.

PrjShrestha commented on shared-libs/transitions/src/transitions/registration.js:
```diff
@@ -185,6 +186,27 @@ const setBirthDate = doc => {
+  const doc = options.doc;
+  const phoneNumber = getPhoneNumber(doc);
+  // By default for a valid phone number SmsParser itself adds a country code if not provided
+  // So if country code is not present in phone_number throw error
```
Rephrased

PrjShrestha commented on shared-libs/validation/src/validation.js:
```diff
@@ -163,6 +163,11 @@ module.exports = {
         })
         .then(result => !result);
     },
+    uniquePhone: (doc, validation) => {
```
Doc update PR [here](https://github.com/medic/cht-docs/pull/1144). Thanks for reminder.

garethbowen commented:
@PrjShrestha It looks like there are some eslint failures causing the build not to complete. I always forget too, but running `grunt eslint` before committing is a quick way to catch these early.

PrjShrestha commented:
@binokaryg @garethbowen Please recheck. Added a small doc entry for uniquephone validation as well.

garethbowen reviewed (approved):
Great work, thanks @PrjShrestha !!

@binokaryg Can you please check that all your comments have been addressed, then it'll be ready to merge.

binokaryg reviewed (approved):
Great!

PrjShrestha commented:
Update: 
Moved logic to create patient or not based upon phone validation to addPatient trigger. 2 benefits:
a. add_phone_number trigger is no longer required.
b. The other option would have been to add a **new validPhone validation in transition** just like we have added uniquePhone. But validating whether a phone is valid or not should be part of the core rather than a choice that should be written in appsettings registration validations.

@tatilepizs This (point b) is why the invalid phone validation was still creating a Patient because we weren't hard exiting addPatient when that validation failed, unlike uniquePhone validation which would hard exit patient creation. Thanks for catching that. 👍 

@dianabarsan Thanks for you help and please kindly review.

dianabarsan commented on api/src/services/report/smsparser.js:
```diff
@@ -209,9 +210,23 @@ const fieldParsers = {
+    const formattedAndValidatedPhone = phoneNumberParser.normalize(config.getAll(), raw);
+    if (formattedAndValidatedPhone) {
+      return formattedAndValidatedPhone;
+    } else {
```
You don't need an else here, because the previous if statement has a return.

dianabarsan commented on shared-libs/transitions/src/transitions/registration.js:
```diff
@@ -40,6 +41,23 @@ const getNameField = (params, prefix) => {
   return defaultNameField;
 };
 
+const getPatientPhoneField = (settings, currentForm) => {
+  // Get the current form for which we are running transition form
+  if(settings && currentForm && settings.forms) {
+    const formDef = settings.forms[currentForm];
+
+    if (formDef && formDef.fields) { 
+      // Get the phone field i.e.the field with type phone_number in the form
+      const phoneField = (Object.keys(formDef.fields).filter(key => formDef.fields[key].type === 'phone_number'));
+      // Return the phone field
+      if (phoneField && phoneField[0]) {
+        return phoneField[0];
+      }
+    }  
+  }
+};
+
```
Please consider early returns in this function to reduce depth, like: 
```suggestion
const getPatientPhoneField = (currentForm) => {
  const formDefinition = utils.getForm(currentForm);
  if (!formDefinition) {
    return;
  }
  
  const phoneField = Object.keys(formDef.fields).find(key => formDef.fields[key].type === 'phone_number');
  return phoneField; 
};
```

I'm also pretty sure we should have a way to get a specific form config, instead of having to go through all settings. 
Found it: https://github.com/medic/cht-core/blob/0e4c653890ce020064f7f0a07fa83c89ae7d06cf/shared-libs/transitions/src/lib/utils.js#L162

dianabarsan commented on shared-libs/transitions/test/unit/transitions/registration.js:
```diff
@@ -2798,7 +2800,9 @@ describe('registration', () => {
-        contactTypeUtils.isPlace.args[0].should.deep.equal([{}, {_id: 'person', patient_id: '56987', type: 'person' }]);
+        contactTypeUtils.isPlace.args[0].should.deep.equal(
+          [{}, { _id: 'person', patient_id: '56987', type: 'person' }]
+        );
```
Please add unit tests to cover this new functionality.

PrjShrestha replied:
Interesting there should be no change here. Think its formatting changes.

PrjShrestha replied:
Was failing the 120 line eslint length check on this test hence Formatted.

dianabarsan replied:
I just picked this line to add the comment. I was referring to adding unit tests in general.

PrjShrestha replied:
Added tests for registration add_patient flow as well as function to get phone field.

dianabarsan commented on tests/integration/sentinel/transitions/registration.spec.js:
```diff
@@ -98,6 +98,145 @@ describe('registration', () => {
+      })
+      .then(() => utils.getDocs(docIds))
+      .then(updated => {
+        console.log(updated);
```
Please remove this debug log.

PrjShrestha replied:
Removed

dianabarsan commented on tests/integration/sentinel/transitions/registration.spec.js:
```diff
@@ -98,6 +98,145 @@ describe('registration', () => {
+      .then(infos => {
+        infos.forEach(info => {
+          chai.expect(info).to.deep.nested.not.include({ 'transitions.registration.ok': false });
+        });
```
Please also check that no patient_id field is added to the main doc.

PrjShrestha replied:
Done.

dianabarsan reviewed (changes requested):
Really cool stuff. Thanks for implementing all the changes. I have some comments but mostly about readability and maintainability. 
Awesome work!

PrjShrestha commented on shared-libs/validation/test/validations.js:
```diff
@@ -248,6 +248,61 @@ describe('validations', () => {
   });
 
+  it('unique phone validation should fail if db query for phone returns doc', () => {
+    clock = sinon.useFakeTimers();
```
Removed. Thanks.

PrjShrestha commented on tests/integration/sentinel/transitions/registration.spec.js:
```diff
@@ -98,6 +98,145 @@ describe('registration', () => {
+        chai.expect(updated[0].fields.phone_number).to.equal(patient_phone);
+        chai.expect(updated[0].patient_id).to.not.be.null;
+      })
+      .then(patients => {
```
Added get patient call and asserted taht its undefined when phone is invalid.

dianabarsan commented on shared-libs/transitions/src/transitions/registration.js:
```diff
@@ -40,6 +41,22 @@ const getNameField = (params, prefix) => {
+    return;
+  }
+
+  if (formDef && formDef.fields) {
```
We can further simplify this like: 
```suggestion
if (!formDef?.fields) {
  return;
}

return Object
  .keys(formDef.fields)
  .find(key => formDef.fields[key].type === 'phone_number');

```

PrjShrestha replied:
Thanks. 
P.S. eslintrc for shared-libs was in 2018 which coudln't recognize the optional chaining syntax (?.) so upgraded it to 2020. 3 or 4 folders are already in 2020.

dianabarsan commented on shared-libs/transitions/src/transitions/registration.js:
```diff
@@ -505,6 +525,14 @@ const addPatient = (options) => {
         }
 
+        if (patientPhoneField && doc.fields[patientPhoneField]) {
+          if (!phoneNumberParser.validate(config.getAll(), (doc.fields[patientPhoneField]))) {
```
To make this whole block more readable, please store `doc.fields[patientPhoneField]` in a variable. 
```suggestion
const patientPhone = doc.fields[patientPhoneField];
if (!phoneNumberParser.validate(config.getAll(), patientPhone)) {
....
  return;
}
patient.phone = patientPhone;
```

PrjShrestha replied:
done.

dianabarsan commented on tests/integration/sentinel/transitions/registration.spec.js:
```diff
@@ -98,6 +98,150 @@ describe('registration', () => {
+
+    return utils
+      .updateSettings(testForm.forms.NP, 'sentinel')
+      .then(utils.saveDocs(docs))
```
```suggestion
 .then(() => utils.saveDocs(docs))
```

dianabarsan commented on tests/integration/sentinel/transitions/registration.spec.js:
```diff
@@ -98,6 +98,150 @@ describe('registration', () => {
   afterEach(() => utils.revertDb(getIds(contacts), true));
 
+  it('should add valid phone to patient doc', () => {
+    const patient_phone = '+9779841123123';
```
Please use camelCase for variables. 
```suggestion
const patientPhone = '+9779841123123';
```

dianabarsan commented on tests/integration/sentinel/transitions/registration.spec.js:
```diff
@@ -98,6 +98,150 @@ describe('registration', () => {
+  });
+
+  it('should not create patient from report doc when provided invalid phone', () => {
+    const patient_phone = '+97796666';
+    const patient_id =uuid();
```
Please use camelCase for variables. 
```suggestion
    const patientPhone = '+97796666';
    const patientId = uuid();
```

dianabarsan commented on tests/integration/sentinel/transitions/registration.spec.js:
```diff
@@ -98,6 +98,150 @@ describe('registration', () => {
+      })
+      .then(() => utils.getDocs(docIds))
+      .then(updated => {
+        chai.expect(updated.patient_id).to.be.undefined;
```
I think that `updated` will be an array here. 
```suggestion
chai.expect(updated[0].patient_id).to.be.undefined;
```

PrjShrestha replied:
Changed thanks.

dianabarsan reviewed (changes requested):
Cool stuff! I spotted some stuff and I'm still missing unit tests updates for the registration transition.

PrjShrestha commented:
Doc update for the phone_number field [here](https://github.com/medic/cht-docs/pull/1152). Also closed PR for add_phone_number trigger as that functionality is done in add_patient.

PrjShrestha commented on tests/integration/sentinel/transitions/registration.spec.js:
```diff
@@ -98,6 +98,150 @@ describe('registration', () => {
+
+    return utils
+      .updateSettings(testForm.forms.NP, 'sentinel')
+      .then(utils.saveDocs(docs))
```
done

PrjShrestha commented on tests/integration/sentinel/transitions/registration.spec.js:
```diff
@@ -98,6 +98,150 @@ describe('registration', () => {
+        return getContactsByReference([newPatientId, 'venus']);
+      })
+      .then(patients => {
+        chai.expect(patients[0]).to.be.undefined;
```
done thanks.

dianabarsan reviewed (approved):
Great stuff @PrjShrestha !

yrimal commented on api/src/services/report/smsparser.js:
```diff
@@ -209,9 +210,22 @@ const fieldParsers = {
+  phone_number: (raw) => {
+    const formattedAndValidatedPhone = phoneNumberParser.normalize(config.getAll(), raw);
+    if (formattedAndValidatedPhone) {
+      return formattedAndValidatedPhone;
```
@PrjShrestha , would you mind adding Nepali digit supporting code for this type as well? Here's the sample [PR](https://github.com/medic/cht-core/pull/8450).

PrjShrestha replied:
Done
